device/cuda/src/utils/make_magnetic_field.cpp

```diff
@@ -12,22 +12,53 @@
 
 // Project include(s).
 #include "traccc/bfield/magnetic_field_types.hpp"
-#include "traccc/definitions/common.hpp"
 
 // System include(s).
 #include <stdexcept>
 
 namespace traccc::cuda {
 
-magnetic_field make_magnetic_field(const magnetic_field& bfield) {
+magnetic_field make_magnetic_field(const magnetic_field& bfield,
+                                   const magnetic_field_storage storage) {
 
-    if (bfield.is<const_bfield_backend_t<scalar>>()) {
-        return magnetic_field{covfie::field<const_bfield_backend_t<scalar>>{
-            bfield.as_field<const_bfield_backend_t<scalar>>()}};
-    } else if (bfield.is<host::inhom_bfield_backend_t<scalar>>()) {
-        return magnetic_field{
-            covfie::field<cuda::inhom_bfield_backend_t<scalar>>(
-                bfield.as_field<host::inhom_bfield_backend_t<scalar>>())};
+    if (bfield.is<const_bfield_backend_t<float>>()) {
+        return magnetic_field{covfie::field<const_bfield_backend_t<float>>{
+            bfield.as_field<const_bfield_backend_t<float>>()}};
+    } else if (bfield.is<const_bfield_backend_t<double>>()) {
+        return magnetic_field{covfie::field<const_bfield_backend_t<double>>{
+            bfield.as_field<const_bfield_backend_t<double>>()}};
+    } else if (bfield.is<host::inhom_bfield_backend_t<float>>()) {
+        // Convenience access to the Covfie field object.
+        const auto& in_field =
+            bfield.as_field<host::inhom_bfield_backend_t<float>>();
+        // At single precision we can use either global or texture memory.
+        if (storage == magnetic_field_storage::global_memory) {
+            return magnetic_field{
+                covfie::field<cuda::inhom_global_bfield_backend_t<float>>(
+                    in_field)};
+        } else if (storage == magnetic_field_storage::texture_memory) {
+            return magnetic_field{
+                covfie::field<cuda::inhom_texture_bfield_backend_t<float>>(
+                    covfie::make_parameter_pack(
+                        in_field.backend().get_configuration(),
+                        in_field.backend()
+                            .get_backend()
+                            .get_backend()
+                            .get_backend()))};
+        } else {
+            throw std::invalid_argument(
+                "Unsupported storage method chosen for inhomogeneous b-field");
+        }
+    } else if (bfield.is<host::inhom_bfield_backend_t<double>>()) {
+        // At double precision we can only use global memory.
+        if (storage == magnetic_field_storage::global_memory) {
+            return magnetic_field{
+                covfie::field<cuda::inhom_global_bfield_backend_t<double>>(
+                    bfield.as_field<host::inhom_bfield_backend_t<double>>())};
+        } else {
+            throw std::invalid_argument(
+                "Unsupported storage method chosen for inhomogeneous b-field");
+        }
```

Comment:
The logic here came out being pretty complicated. 😦 But after thinking about it for a bit, this still seemed the best way of implementing it.

I'm open to others having better ideas though.